src/v/datalake/tests/record_schema_resolver_test.cc

```diff
@@ -261,3 +263,191 @@ TEST_F(RecordSchemaResolverTest, TestSchemaRegistryError) {
+    counting_store(
+      schema::fake_registry& registry,
+      std::unordered_map<pandaproxy::schema_registry::schema_id, size_t>&
+        counts)
+      : registry(registry)
+      , counts(counts) {}
+
+    ss::future<pandaproxy::schema_registry::subject_schema> get_subject_schema(
+      pandaproxy::schema_registry::subject sub,
+      std::optional<pandaproxy::schema_registry::schema_version> version,
+      pandaproxy::schema_registry::include_deleted inc_dec) final {
+        auto* getter = co_await registry.getter();
+        co_return co_await getter->get_subject_schema(sub, version, inc_dec);
+    }
+
+    ss::future<pandaproxy::schema_registry::canonical_schema_definition>
+    get_schema_definition(pandaproxy::schema_registry::schema_id id) final {
+        counts[id] += 1;
+        auto* getter = co_await registry.getter();
+        co_return co_await getter->get_schema_definition(id);
+    }
+
+    ss::future<
+      std::optional<pandaproxy::schema_registry::canonical_schema_definition>>
+    maybe_get_schema_definition(
+      pandaproxy::schema_registry::schema_id id) final {
+        counts[id] += 1;
+        auto* getter = co_await registry.getter();
+        co_return co_await getter->maybe_get_schema_definition(id);
+    }
+
+    schema::fake_registry& registry;
+    std::unordered_map<pandaproxy::schema_registry::schema_id, size_t>& counts;
+};
+
+class counting_registry : public schema::registry {
+public:
+    bool is_enabled() const override { return true; };
+
+    ss::future<pandaproxy::schema_registry::schema_getter*>
+    getter() const override {
+        co_return &_store;
+    }
+    ss::future<pandaproxy::schema_registry::canonical_schema_definition>
+    get_schema_definition(
+      pandaproxy::schema_registry::schema_id id) const override {
+        return _store.get_schema_definition(id);
+    }
+
+    ss::future<pandaproxy::schema_registry::subject_schema> get_subject_schema(
+      pandaproxy::schema_registry::subject sub,
+      std::optional<pandaproxy::schema_registry::schema_version> version)
+      const override {
+        return _registry.get_subject_schema(sub, version);
+    }
+
+    ss::future<pandaproxy::schema_registry::schema_id> create_schema(
+      pandaproxy::schema_registry::unparsed_schema unparsed) override {
+        return _registry.create_schema(std::move(unparsed));
+    }
+
+    const std::vector<pandaproxy::schema_registry::subject_schema>& get_all() {
+        return _registry.get_all();
+    }
+
+    size_t get_count(pandaproxy::schema_registry::schema_id id) {
+        return _counts[id];
+    }
+
+    void reset_counts() { _counts.clear(); }
+
+private:
+    schema::fake_registry _registry{};
+    std::unordered_map<pandaproxy::schema_registry::schema_id, size_t>
+      _counts{};
+    mutable counting_store _store{_registry, _counts};
+};
+
+schema_cache_t make_schema_cache() {
+    return schema_cache_t{schema_cache_t::config{10, 4}};
```

Comment:
Can you also add a test case where a schema is evicted from the cache, triggering another count? And maybe also validate accessing a shared schema after it's been evicted?

Reply:
Yeah, can do. Originally this PR also included https://github.com/redpanda-data/redpanda/pull/24514 so a lot of the testing was more focused on the kv cache. However, it wouldn't hurt to repeat a lot of that testing in the datalake layer as well.